File: src/converter/plugins/VarianceModifierPlugin.ts
import {ConverterPlugin} from "../plugin.js";
import ts, {Node} from "typescript";
import {ConverterContext} from "../context.js";
import {Render} from "../render.js";
import {GeneratedFile} from "../generated.js";
import {VarianceModifier} from "../varianceModifier.js";

export const varianceModifierServiceKey = Symbol()

export class VarianceModifierService {
    constructor(private readonly varianceModifiers: VarianceModifier[]) {
    }

    resolveVarianceModifier(
        node: Node,
        context: ConverterContext,
    ): string | null {
        for (const varianceModifier of this.varianceModifiers) {
            const result = varianceModifier(node, context)

            if (result !== null) return result
        }

        return null
    }
}

export class VarianceModifierPlugin implements ConverterPlugin {
    private readonly varianceModifierService: VarianceModifierService;

    constructor(varianceModifiers: VarianceModifier[]) {
        this.varianceModifierService = new VarianceModifierService(varianceModifiers)
    }

    generate(): GeneratedFile[] {
        return [];
    }

    render(node: Node, context: ConverterContext, next: Render): string | null {
        return null;
    }

    traverse(node: ts.Node, context: ConverterContext): void {
    }

    setup(context: ConverterContext): void {
        context.registerService(varianceModifierServiceKey, this.varianceModifierService)
    }
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/VarianceModifierPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.*
import js.array.ReadonlyArray
import js.symbol.Symbol
import typescript.Node

val varianceModifierServiceKey = Symbol()

class VarianceModifierService(
    private val varianceModifiers: ReadonlyArray<VarianceModifier<Node>>
) {
    fun resolveVarianceModifier(
        node: Node,
        context: ConverterContext,
    ): String? {
        for (varianceModifier in varianceModifiers) {
            val result = varianceModifier(node, context)

            if (result != null) return result
        }

        return null
    }
}

class VarianceModifierPlugin(varianceModifiers: ReadonlyArray<VarianceModifier<Node>>) : ConverterPlugin<Node> {
    private val varianceModifierService = VarianceModifierService(varianceModifiers);

    override fun setup(context: ConverterContext) {
        context.registerService(varianceModifierServiceKey, varianceModifierService)
    }

    override fun traverse(node: Node, context: ConverterContext) = Unit

    override fun render(node: Node, context: ConverterContext, next: Render<Node>) = null

    override fun generate(context: ConverterContext, render: Render<Node>) = emptyArray<GeneratedFile>()
}
